js/source/video_source.js

```diff
@@ -146,16 +146,13 @@ VideoSource.prototype = util.inherit(Evented, /** @lends VideoSource.prototype *
         var gl = painter.gl;
         if (!this.tile.texture) {
             this.tile.texture = gl.createTexture();
-            gl.bindTexture(gl.TEXTURE_2D, this.tile.texture);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
-            gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
-            gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, this.video);
-        } else {
-            gl.bindTexture(gl.TEXTURE_2D, this.tile.texture);
-            gl.texSubImage2D(gl.TEXTURE_2D, 0, 0, 0, gl.RGBA, gl.UNSIGNED_BYTE, this.video);
         }
+        gl.bindTexture(gl.TEXTURE_2D, this.tile.texture);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA, gl.RGBA, gl.UNSIGNED_BYTE, this.video);
```

Comment:
`texImage2D` allocates new memory and `texSubImage2D` doesn't. It just overwrites the previously allocated memory. Since we're calling this on every frame we should keep using `texSubImage2D` for updates. Otherwise the framerate starts dropping over time.

This difference isn't documented clearly anywhere so it's pretty much impossible to know unless you've been bitten by it already. It would probably be a good idea to add a comment here that explains why `texSubImage2D` is used.